src/overlays/actors/ovl_Obj_Ice_Poly/z_obj_ice_poly.c

```diff
@@ -27,15 +31,156 @@ const ActorInit Obj_Ice_Poly_InitVars = {
+        this->actionFunc = ObjIcePoly_Melt;
+    } else if (this->actor.parent != NULL) {
+        this->actor.parent->freezeTimer = 40;
+        CollisionCheck_SetAT(globalCtx, &globalCtx->colChkCtx, &this->colliderIce.base);
+        CollisionCheck_SetAC(globalCtx, &globalCtx->colChkCtx, &this->colliderIce.base);
+        CollisionCheck_SetOC(globalCtx, &globalCtx->colChkCtx, &this->colliderIce.base);
+        CollisionCheck_SetAC(globalCtx, &globalCtx->colChkCtx, &this->colliderHard.base);
+    } else {
+        Actor_Kill(&this->actor);
+    }
+    pos.x = this->actor.posRot.pos.x +
+            this->actor.scale.x * (Math_Rand_S16Offset(15, 15) * (Math_Rand_ZeroOne() < 0.5f ? -1 : 1));
+    pos.y = this->actor.posRot.pos.y + this->actor.scale.y * Math_Rand_S16Offset(10, 90);
+    pos.z = this->actor.posRot.pos.z +
+            this->actor.scale.z * (Math_Rand_S16Offset(15, 15) * (Math_Rand_ZeroOne() < 0.5f ? -1 : 1));
+    if ((globalCtx->gameplayFrames % 7) == 0) {
+        EffectSsKiraKira_SpawnDispersed(globalCtx, &pos, &zeroVec, &zeroVec, &sColorWhite, &sColorGray, 2000, 5);
+    }
+}
+
+void ObjIcePoly_Melt(ObjIcePoly* this, GlobalContext* globalCtx) {
+    Vec3f accel;
+    Vec3f vel;
+    Vec3f pos;
+    s32 i;
+
+    accel.x = 0.0f;
+    accel.y = this->actor.scale.y;
+    accel.z = 0.0f;
+    vel.x = 0.0f;
+    vel.y = this->actor.scale.y;
+    vel.z = 0.0f;
+
+    for (i = 0; i < 2; i++) {
+        pos.x = this->actor.posRot.pos.x +
+                this->actor.scale.x * (Math_Rand_S16Offset(20, 20) * (Math_Rand_ZeroOne() < 0.5f ? -1 : 1));
+        pos.y = this->actor.posRot.pos.y + this->actor.scale.y * Math_Rand_ZeroOne() * 50.0f;
+        pos.z = this->actor.posRot.pos.z +
+                this->actor.scale.x * (Math_Rand_S16Offset(20, 20) * (Math_Rand_ZeroOne() < 0.5f ? -1 : 1));
+        func_8002829C(globalCtx, &pos, &vel, &accel, &sColorWhite, &sColorGray,
+                      Math_Rand_S16Offset(0x15E, 0x64) * this->actor.scale.x, this->actor.scale.x * 20.0f);
+    }
+    if (this->meltTimer < 0) {
+        if (this->actor.parent != NULL) {
+            this->actor.parent->freezeTimer = 40;
+        }
+        this->meltTimer++;
+        if (this->meltTimer == 0) {
+            this->meltTimer = 40;
+            Audio_PlayActorSound2(&this->actor, NA_SE_EV_ICE_MELT);
+        }
+    } else {
+        if (this->meltTimer != 0) {
+            this->meltTimer--;
+        }
+        this->actor.scale.y = sScale[this->actor.params] * (0.5f + (this->meltTimer * 0.0125f));
+        this->alpha -= 6;
+        if (this->meltTimer == 0) {
+            Actor_Kill(&this->actor);
+        }
+    }
+}
+
+void ObjIcePoly_Update(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    ObjIcePoly* this = THIS;
+
+    this->actionFunc(this, globalCtx);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Obj_Ice_Poly/func_80B94470.s")
+void ObjIcePoly_Draw(ObjIcePoly* thisx, GlobalContext* globalCtx) {
```

Comment:
Also need to change this.
```suggestion
void ObjIcePoly_Draw(Actor* thisx, GlobalContext* globalCtx) {
```